### PR TITLE
fix: support batch broadcasting in JoinImageWithAlpha node

### LINKED ISSUE #12580: Join Image With Alpha node does not work with image batching
### Custom Node Testing

- [x] I have tried disabling custom nodes and the issue persists (see [how to disable custom nodes](https://docs.comfy.org/troubleshooting/custom-node-issues#step-1%3A-test-with-all-custom-nodes-disabled) if you need help)

### Expected Behavior

When passing a batch of images to the `Join Image with Alpha` node, it should return a batch of images.

### Actual Behavior

The `Join Image with Alpha` node always returns a single image.

### Steps to Reproduce

1. Run this workflow: [Bug.json](https://github.com/user-attachments/files/25470595/Bug.json)
2. Notice that it only returns a single image.

### Debug Logs

```powershell
[START] Security scan
[ComfyUI-Manager] Using uv as Python module for pip operations.
Using Python 3.10.16 environment at: env
[DONE] Security scan
## ComfyUI-Manager: installing dependencies done.
** ComfyUI startup time: 2026-02-21 16:00:03.213
** Platform: Linux
** Python version: 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
** Python executable: /home/pauan/.config/pinokio/api/comfy.git/app/env/bin/python
** ComfyUI Path: /home/pauan/.config/pinokio/api/comfy.git/app
** ComfyUI Base Folder Path: /home/pauan/.config/pinokio/api/comfy.git/app
** User directory: /home/pauan/.config/pinokio/api/comfy.git/app/user
** ComfyUI-Manager config path: /home/pauan/.config/pinokio/api/comfy.git/app/user/default/ComfyUI-Manager/config.ini
** Log path: /home/pauan/.config/pinokio/api/comfy.git/app/user/comfyui.log
Using Python 3.10.16 environment at: env
Using Python 3.10.16 environment at: env

Prestartup times for custom nodes:
   0.2 seconds: /home/pauan/.config/pinokio/api/comfy.git/app/custom_nodes/ComfyUI-Manager

WARNING: You need pytorch with cu130 or higher to use optimized CUDA operations.
Checkpoint files will always be loaded safely.
Total VRAM 15838 MB, total RAM 96175 MB
pytorch version: 2.7.0+cu128
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 5070 Ti : cudaMallocAsync
Using async weight offloading with 2 streams
Enabled pinned memory 91365.0
Using pytorch attention
Python version: 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
ComfyUI version: 0.14.1
ComfyUI frontend version: 1.39.14
[Prompt Server] web root: /home/pauan/.config/pinokio/api/comfy.git/app/env/lib/python3.10/site-packages/comfyui_frontend_package/static
### Loading: ComfyUI-Manager (V3.37.2)
[ComfyUI-Manager] network_mode: public
### ComfyUI Version: v0.14.1-24-g7591d781 | Released on '2026-02-21'
```

### Other

This is happening because the images are batched, but the mask is a single value.

However, this shouldn't be a problem, because every other node in ComfyUI is able to correctly combine a batch with a single value.

For example, when passing a batch of images to the `ImageCompositeMasked` node, it returns a batch of images, even if the `source` and `mask` are a single value.

So the behavior of `Join Image with Alpha` is different from every other node in ComfyUI.

In addition, the `RepeatMaskBatch` node doesn't exist, so there is no easy way for the user to fix the incorrect behavior of `Join Image with Alpha`.

### PR DESCRIPTION
**Pull request by Abdulrehman-PIAIC80387**
## Summary

Fixes #12580

### Problem

The `Join Image with Alpha` node only returns a single image when given a batch of images with a single alpha mask.

**Root cause:** In `comfy_extras/nodes_compositing.py` line 205, the batch size was calculated using `min(len(image), len(alpha))`. When a user passes a batch of 4 images with a single alpha mask, `min(4, 1) = 1`, so only one image is produced.

This is inconsistent with every other compositing node in ComfyUI (e.g., `ImageCompositeMasked`), which correctly broadcasts the shorter input to match the longer one.

### Fix

Changed `comfy_extras/nodes_compositing.py` to use `max()` instead of `min()` for batch size, with modulo indexing (`i % len(...)`) to repeat the shorter input:

```python
# Before
batch_size = min(len(image), len(alpha))
...
out_images.append(torch.cat((image[i][:,:,:3], alpha[i].unsqueeze(2)), dim=2))

# After
batch_size = max(len(image), len(alpha))
...
out_images.append(torch.cat((image[i % len(image)][:,:,:3], alpha[i % len(alpha)].unsqueeze(2)), dim=2))
```

This handles all cases correctly:
- **4 images + 1 mask** → 4 output images (mask repeats)
- **1 image + 4 masks** → 4 output images (image repeats)
- **4 images + 4 masks** → 4 output images (1-to-1)
- **1 image + 1 mask** → 1 output image (unchanged behavior)

### CODE REVIEW

**Comment by alexisrolland** [comment]
@Abdulrehman-PIAIC80387 thank you for your contribution, this looks good.

Could you please address the error from the CI pipeline and update your PR?

```
Error: AI agent Co-authored-by trailers detected in PR commits.
The following commits contain Co-authored-by trailers from AI coding agents:
  bb5c3378: Co-Authored-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

These trailers should be removed before merging.
To fix, rewrite the commit messages with:
  git rebase -i d0f0b15cf5d1fbff67390c8d90ec8654c2582f7a

and remove the Co-authored-by lines, then force-push your branch.
```

Thanks

**Comment by Abdulrehman-PIAIC80387** [comment]
@alexisrolland Done — I removed the Co-authored-by trailer from the commit message and force-pushed. Thanks for the review!

**Review by alexisrolland — APPROVED** [review]
Tested and it works.

Image


Image


{B946D84F-EF1C-4770-A48D-8FF402DC9974}

**Comment by kijai** [comment]
Hey, thank you for raising the issue and for the suggested fix.

However I would rather do this using the existing `comfy.utils.repeat_to_batch_size` helper for the purpose, which handles the error case here correctly already, so I made separate PR for that: https://github.com/Comfy-Org/ComfyUI/pull/13686

**Comment by alexisrolland** [comment]
Agree with @kijai the `comfy.utils.repeat_to_batch_size` is more elegant so I'll close this one. Thanks for your contribution @Abdulrehman-PIAIC80387 !
